### DISCUSSION TITLE
Clarify Contribution Policy

### DISCUSSION THREAD

**Discussion by divanise** [enisdenjo/graphql-ws#537] [Q&A] [answered]
Hi!

First, let me add my thanks to everyone else's for the great project!

For background, #333 is a problem for me and I'm not entirely happy with the custom `makeServer` workaround suggested there. I also completely understand that you're reluctant to change the existing behaviour, so I'm looking into making them myself in a fork.

As a first step, [I've added a number of tests](https://github.com/lazarentertainment/graphql-ws/tree/expanded-tests) to document the existing behaviour of how errors are transported from the `subscribe`/`resolve` functions to the client. I'd like to contribute the tests back, but it's not at all clear to me how to start that process.

### Story

As a developer who forked the repo,

I want to know how to discuss contributing my changes back upstream,

So that others users of the project can benefit.

### Acceptance criteria

- The contribution guidelines have a section dedicated to what to do if you already have code you'd like to discuss contributing.
  - If that involves directly opening a pull request, explicitly mention that that is OK. (A lot of projects don't appreciate drive-by pull requests.)
  - If an issue should be filed, either provide a new template for contributions or clarify whether a contribution should use the bug report or user story templates.

**Comment by enisdenjo** [comment, marked as answer]
Hey there, thanks for the contribution interest!

Regarding #333, it is an issue with graphql-js and not graphql-ws itself. See https://github.com/graphql/graphql-js/issues/1539 and https://github.com/graphql/graphql-js/issues/4001. Also, the reason why graphql-ws doesn't offer built-in handling of internal errors is because there's no one-glove-fits-all. Errors are important and should be handled with care, exactly to the needs of the user - not the library author.

Regarding contributions, aside from what's stated in the [contributing doc](https://github.com/enisdenjo/graphql-ws/blob/master/CONTRIBUTING.md), I have no preference. You're free to open a PR or an issue, or both. IMHO, if you have a PR handy, I don't think there's a need to additionally have an issue open. All the necessary discussion can take place there.

**Reply by divanise** [reply]
> You're free to open a PR or an issue, or both.

Done, #538.

> Regarding https://github.com/enisdenjo/graphql-ws/discussions/333, it is an issue with graphql-js and not graphql-ws itself.

I understand your position, and I don't even entirely disagree with it. From my project's perspective, though, changing our subscription library broke our application, and this looks like the easier place for us to make a fix.

> Errors are important and should be handled with care, exactly to the needs of the user - not the library author.

I totally agree, which is why it's frustrating that there's this one single error code path that we sometimes hit where we can't override *either* library's approach to error handling. 😄

**Reply by enisdenjo** [reply]
Also, if you don't want `subscribe` to throw, you can simply wrap graphql-js subscribe with your own and handle the error however you want and provide it in the config.

**Reply by divanise** [reply]
Yup, I've taken that approach in [this commit](https://github.com/lazarentertainment/graphql-ws/commit/8f0b8a7ead6619992a86fb2d41378cd020dd341f), but the semantics of wrapping an async iterator such that the `return` method on the iterator is properly handled are far from simple.

My preferred approach is in [this commit](https://github.com/lazarentertainment/graphql-ws/commit/c490b4e3879173e2b80fc95bf127c856c7a1a0f4), which adds a new callback that allows the client of the library to override the default behaviour of closing the socket.

**Reply by enisdenjo** [reply]
Thanks for working on that! But I am not keen on offering built-in handlers for such cases because of aforementioned reasons.

Nevertheless, a custom subscribe is quite simple - I've written a recipe showcasing how someone may [implement a custom subscribe for graceful handling thrown errors](https://the-guild.dev/graphql/ws/recipes#ws-server-usage-with-custom-subscribe-method-that-gracefully-handles-thrown-errors) with just 15 lines of code.

**Reply by divanise** [reply]
Those fifteen lines of code don't properly handle iterator cancellation. That's the first approach I tried, and when I used a version of that wrapper (without the try/catch block, so I was just testing that the wrapping worked) on the full test suite for this project, several of the cancellation tests failed -- at least some of the failures were because the wrapper in your recipe doesn't return true from `isAsyncGenerator` when wrapping a generator.

And you're very welcome! I was hoping that the new callback I added would be interesting to you, but I understand where you're coming from and respect your decision.

**Reply by enisdenjo** [reply]
> when I used a version of that wrapper (without the try/catch block, so I was just testing that the wrapping worked) on the full test suite for this project, several of the cancellation tests failed

Hmm, are you sure you did it as per the recipe? I did the recipe change internally and run the full test suite with all tests passing.

**Reply by divanise** [reply]
On the one hand, I skimmed over your link a little too quickly, sorry about that. My original test involved creating a new iterator to wrap the old one rather than modifying the old one in place.

However, using the same approach as your recipe on v5.15.0 still fails one test for me:

```diff
diff --git a/src/server.ts b/src/server.ts
index ff116a7..6771a7e 100644
--- a/src/server.ts
+++ b/src/server.ts
@@ -800,7 +800,14 @@ export function makeServer<
                 execArgs,
                 operationResult,
               );
-              if (maybeResult) operationResult = maybeResult;
+              if (maybeResult) {
+                operationResult = maybeResult;
+              } else if ('next' in operationResult) {
+                const originalNext = operationResult.next;
+                operationResult.next = async () => {
+                  return await originalNext();
+                };
+              }
 
               if (isAsyncIterable(operationResult)) {
                 /** multiple emitted results */
```

```
 FAIL  src/__tests__/server.ts (11.823 s)
  ● Subscribe › should execute the live query, "next" multiple results and then "complete"

    Timeout - Async callback was not invoked within the 5000 ms timeout specified by jest.setTimeout.Error:
```

**Reply by enisdenjo** [reply]
That is because the `this` gets polluted there. You have to bind the original next method.

```diff
- const originalNext = operationResult.next;
+ const originalNext = operationResult.next.bind(operationResult);
```

**Reply by divanise** [reply]
Ah, got it. Thanks!
